<a href="https://colab.research.google.com/github/nikitask14/cross-site-heart-disease-generalisation/blob/main/NN_Research_Capstone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###**cross-site-heart-disease-generalisation**
**Stage1: Understanding the dataset**

In [1]:
import pandas as pd
import numpy as np
cleveland_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"
cleveland = pd.read_csv(cleveland_url, header = None)

hungarian_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.hungarian.data"
hungary = pd.read_csv(hungarian_url, header = None)

switzerland_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.switzerland.data"
switzerland = pd.read_csv(switzerland_url, header = None)

va_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.va.data"
va = pd.read_csv(va_url, header = None)


header=None

pandas had no column names available, so it automatically labelled the 14 columns: 0, 1, 2, 3, ... 13

In [2]:
cleveland.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       303 non-null    float64
 1   1       303 non-null    float64
 2   2       303 non-null    float64
 3   3       303 non-null    float64
 4   4       303 non-null    float64
 5   5       303 non-null    float64
 6   6       303 non-null    float64
 7   7       303 non-null    float64
 8   8       303 non-null    float64
 9   9       303 non-null    float64
 10  10      303 non-null    float64
 11  11      303 non-null    object 
 12  12      303 non-null    object 
 13  13      303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


In [3]:
hungary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       294 non-null    int64  
 1   1       294 non-null    int64  
 2   2       294 non-null    int64  
 3   3       294 non-null    object 
 4   4       294 non-null    object 
 5   5       294 non-null    object 
 6   6       294 non-null    object 
 7   7       294 non-null    object 
 8   8       294 non-null    object 
 9   9       294 non-null    float64
 10  10      294 non-null    object 
 11  11      294 non-null    object 
 12  12      294 non-null    object 
 13  13      294 non-null    int64  
dtypes: float64(1), int64(4), object(9)
memory usage: 32.3+ KB


In [4]:
switzerland.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       123 non-null    int64 
 1   1       123 non-null    int64 
 2   2       123 non-null    int64 
 3   3       123 non-null    object
 4   4       123 non-null    int64 
 5   5       123 non-null    object
 6   6       123 non-null    object
 7   7       123 non-null    object
 8   8       123 non-null    object
 9   9       123 non-null    object
 10  10      123 non-null    object
 11  11      123 non-null    object
 12  12      123 non-null    object
 13  13      123 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 13.6+ KB


In [5]:
va.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       200 non-null    int64 
 1   1       200 non-null    int64 
 2   2       200 non-null    int64 
 3   3       200 non-null    object
 4   4       200 non-null    object
 5   5       200 non-null    object
 6   6       200 non-null    int64 
 7   7       200 non-null    object
 8   8       200 non-null    object
 9   9       200 non-null    object
 10  10      200 non-null    object
 11  11      200 non-null    object
 12  12      200 non-null    object
 13  13      200 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 22.0+ KB


**info()** only tells us pandas did not recognize any entries as NaN. But this raw dataset may use a symbol such as **?** to represent missing data.

Pandas would treat **?** as an ordinary string, which could explain why many columns in the 4 datasets became object.

In [6]:
cleveland.head()

,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


**Note:**
1. The columns are just numbered 0 through 13. That is because the raw file contained no column names.
2. Before we inspect distributions, missing values, target, or anything else, we need to attach the correct variable names and understand them.

In [7]:
column_names = ["age","sex", "cp", "trestbps", "chol", "fbs","restecg","thalach", "exang", "oldpeak", "slope", "ca", "thal","num"]
cleveland.columns = column_names
hungary.columns = column_names
switzerland.columns = column_names
va.columns = column_names



In [8]:
cleveland.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


In [9]:
hungary.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,28,1,2,130,132,0,2,185,0,0.0,?,?,?,0
1,29,1,2,120,243,0,0,160,0,0.0,?,?,?,0
2,29,1,2,140,?,0,0,170,0,0.0,?,?,?,0
3,30,0,1,170,237,0,1,170,0,0.0,?,?,6,0
4,31,0,2,100,219,0,1,150,0,0.0,?,?,?,0


In [10]:
switzerland.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,32,1,1,95,0,?,0,127,0,.7,1,?,?,1
1,34,1,4,115,0,?,?,154,0,.2,1,?,?,1
2,35,1,4,?,0,?,0,130,1,?,?,?,7,3
3,36,1,4,110,0,?,0,125,1,1,2,?,6,1
4,38,0,4,105,0,?,0,166,0,2.8,1,?,?,2


In [11]:
va.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63,1,4,140,260,0,1,112,1,3,2,?,?,2
1,44,1,4,130,209,0,1,127,0,0,?,?,?,0
2,60,1,4,132,218,0,1,140,1,1.5,3,?,?,2
3,55,1,4,142,228,0,1,149,1,2.5,1,?,?,1
4,66,1,3,110,213,1,2,99,1,1.3,2,?,?,0


**Compare the size of each site**

This matters because unequal site sizes are one possible source of heterogeneity and will matter later when we think about pooled training.

In [12]:
print(cleveland.shape)
print(hungary.shape)
print(switzerland.shape)
print(va.shape)

(303, 14)
(294, 14)
(123, 14)
(200, 14)




```
Cleveland     303 rows
Hungary       294 rows
Switzerland   123 rows
VA Long Beach 200 rows
```



All four have the same 14-column structure, but the number of patient records differs by site.

That is our first concrete sign of heterogeneity:

same schema
≠
same dataset size

In [13]:
cleveland.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        303 non-null    object 
 12  thal      303 non-null    object 
 13  num       303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


Both **ca** and **thal** are supposed to contain numeric codes, yet pandas has stored them as object.

That suggests:

something non-numeric may be present in those columns

In [14]:
cleveland["ca"].unique()

array(['0.0', '3.0', '2.0', '1.0', '?'], dtype=object)

1. The ? is being used as a missing-value marker in the raw file.

2. Because ? is text, pandas cannot treat the whole column as numeric, so it stores the column as object.

3. This also corrects our earlier observation: 303 non-null

4. It did not mean “no missing data.” It only meant pandas had not yet recognised ? as missing.

In [15]:
cleveland["thal"].unique()

array(['6.0', '3.0', '7.0', '?'], dtype=object)

In [16]:
(cleveland["ca"] == "?").sum()

np.int64(4)

In [17]:
(cleveland["thal"] == '?').sum()

np.int64(2)

In [18]:
# clean way to do it
(cleveland == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


**Part B: Inspect Datatype of Hungarian Dataset**

In [19]:
hungary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       294 non-null    int64  
 1   sex       294 non-null    int64  
 2   cp        294 non-null    int64  
 3   trestbps  294 non-null    object 
 4   chol      294 non-null    object 
 5   fbs       294 non-null    object 
 6   restecg   294 non-null    object 
 7   thalach   294 non-null    object 
 8   exang     294 non-null    object 
 9   oldpeak   294 non-null    float64
 10  slope     294 non-null    object 
 11  ca        294 non-null    object 
 12  thal      294 non-null    object 
 13  num       294 non-null    int64  
dtypes: float64(1), int64(4), object(9)
memory usage: 32.3+ KB


We know that Hungary dataset has numerical values but columns 3,4,5,6,7,8,10,11 and 12 have object datatype. We should inspect for missing values.

? is only one possible missing-value marker. A raw dataset could use things like:


```
?
NA
N/A
-
unknown
blank cell
999
-1

```



In [20]:
for column in hungary.columns:
  print(column, hungary[column].unique())

age [28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51
 52 53 54 55 56 57 58 59 60 61 62 63 65 66]
sex [1 0]
cp [2 1 3 4]
trestbps ['130' '120' '140' '170' '100' '105' '110' '125' '150' '98' '112' '145'
 '190' '160' '115' '142' '180' '132' '135' '?' '108' '124' '113' '122'
 '92' '118' '106' '200' '138' '136' '128' '155']
chol ['132' '243' '?' '237' '219' '198' '225' '254' '298' '161' '214' '220'
 '160' '167' '308' '264' '166' '340' '209' '260' '211' '173' '283' '194'
 '223' '315' '275' '297' '292' '182' '200' '204' '241' '339' '147' '273'
 '307' '289' '215' '281' '250' '184' '245' '291' '295' '269' '196' '268'
 '228' '358' '201' '249' '266' '186' '207' '218' '412' '224' '238' '230'
 '163' '240' '280' '257' '263' '276' '284' '195' '227' '253' '187' '202'
 '328' '168' '216' '129' '190' '188' '179' '210' '272' '180' '100' '259'
 '468' '274' '320' '221' '309' '312' '171' '208' '246' '305' '217' '365'
 '344' '394' '256' '326' '277' '270' '229' '85' '347' '251' '222' '2

This systematic inspection tells us several useful things.

For Hungary, the suspicious non-numeric marker we actually observe is consistently:

?

In [21]:
(hungary == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,1
chol,23
fbs,8
restecg,1
thalach,1
exang,1
oldpeak,0


1. slope → 190 missing out of 294
2. ca    → 291 missing out of 294
3. thal  → 266 missing out of 294

**Part C: Inspect Datatype of Switzerland Dataset**

In [22]:
switzerland.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   age       123 non-null    int64 
 1   sex       123 non-null    int64 
 2   cp        123 non-null    int64 
 3   trestbps  123 non-null    object
 4   chol      123 non-null    int64 
 5   fbs       123 non-null    object
 6   restecg   123 non-null    object
 7   thalach   123 non-null    object
 8   exang     123 non-null    object
 9   oldpeak   123 non-null    object
 10  slope     123 non-null    object
 11  ca        123 non-null    object
 12  thal      123 non-null    object
 13  num       123 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 13.6+ KB


In [23]:
for column in switzerland.columns:
  print(column, switzerland[column].unique())

age [32 34 35 36 38 40 41 42 43 45 46 47 48 50 51 52 53 54 55 56 57 58 59 60
 61 62 63 64 65 66 67 68 69 70 72 73 74]
sex [1 0]
cp [1 4 3 2]
trestbps ['95' '115' '?' '110' '105' '100' '135' '150' '125' '145' '140' '155'
 '160' '120' '130' '165' '80' '180' '170' '200' '185']
chol [0]
fbs ['?' '0' '1']
restecg ['0' '?' '1' '2']
thalach ['127' '154' '130' '125' '166' '156' '179' '128' '150' '120' '144' '176'
 '99' '122' '145' '140' '138' '133' '113' '118' '149' '124' '110' '139'
 '92' '104' '170' '163' '60' '126' '82' '95' '115' '135' '141' '155' '83'
 '97' '98' '100' '148' '103' '121' '131' '182' '105' '175' '94' '119'
 '143' '63' '70' '77' '117' '123' '134' '72' '78' '109' '86' '114' '93'
 '67' '90' '108' '136' '?' '157']
exang ['0' '1' '?']
oldpeak ['.7' '.2' '?' '1' '2.8' '0' '-1.1' '1.6' '-1.5' '1.5' '2' '.5' '-.1'
 '-2.6' '2.1' '-.7' '2.2' '3' '.1' '.3' '-2' '-1' '1.8' '1.4' '2.6' '.9'
 '2.4' '1.1' '.4' '2.5' '1.7' '-.8' '-.5' '-.9' '3.7' '1.3']
slope ['1' '?' '2' '3']
ca ['?' '1' '

In [24]:
(switzerland == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,2
chol,0
fbs,75
restecg,1
thalach,1
exang,1
oldpeak,6


1. 123 patients
2. explicit "?" missingness in several variables
3. ca missing for 118/123
4. thal missing for 52/123
5. fbs missing for 75/123
6. AND cholesterol appears unavailable for the entire site

**Part D: Inspect Datatype of VA Dataset**


In [25]:
va.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   age       200 non-null    int64 
 1   sex       200 non-null    int64 
 2   cp        200 non-null    int64 
 3   trestbps  200 non-null    object
 4   chol      200 non-null    object
 5   fbs       200 non-null    object
 6   restecg   200 non-null    int64 
 7   thalach   200 non-null    object
 8   exang     200 non-null    object
 9   oldpeak   200 non-null    object
 10  slope     200 non-null    object
 11  ca        200 non-null    object
 12  thal      200 non-null    object
 13  num       200 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 22.0+ KB


In [26]:
for column in va.columns:
  print(column, va[column].unique())

age [63 44 60 55 66 65 56 59 62 57 46 58 64 74 52 69 51 54 77 61 40 41 42 53
 68 67 72 75 49 35 43 48 50 45 76 70 71 38 37]
sex [1 0]
cp [4 3 2 1]
trestbps ['140' '130' '132' '142' '110' '120' '150' '180' '160' '126' '?' '128'
 '170' '152' '116' '124' '0' '122' '144' '154' '125' '104' '136' '134'
 '138' '178' '146' '135' '158' '106' '112' '102' '96' '172' '155' '156'
 '118' '100' '190' '114' '127']
chol ['260' '209' '218' '228' '213' '0' '236' '267' '166' '220' '177' '186'
 '100' '171' '230' '281' '203' '277' '233' '240' '153' '224' '316' '311'
 '270' '217' '214' '252' '339' '216' '276' '458' '241' '384' '297' '248'
 '308' '208' '227' '210' '245' '225' '198' '195' '161' '258' '235' '305'
 '223' '282' '349' '?' '160' '312' '283' '142' '211' '306' '222' '202'
 '197' '204' '274' '192' '298' '272' '200' '261' '181' '221' '175' '219'
 '310' '232' '273' '182' '292' '289' '193' '170' '369' '173' '271' '244'
 '285' '243' '237' '165' '287' '256' '264' '226' '207' '284' '337' '254'
 '300' '333' 

In [27]:
(va == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,56
chol,7
fbs,7
restecg,0
thalach,53
exang,53
oldpeak,56


## First-Pass Multi-Site Dataset Inspection

The four processed UCI Heart Disease site datasets were loaded separately:

* Cleveland
* Hungary
* Switzerland
* VA Long Beach

All four datasets use the same 14-variable schema, but they differ in both sample size and data completeness.

Initial inspection included:

* dataset shapes,
* column names and data types,
* unique-value inspection,
* identification of `?` as an explicit missing-value marker,
* column-wise missing-value counts for each site.

An important observation is that sharing the same schema does **not** mean the sites contain equally complete data. Hungary, Switzerland, and VA Long Beach have substantial missingness in several variables, particularly `ca`, `thal`, and `slope`.

The Switzerland cholesterol column also requires further investigation because all observed values are `0`, suggesting that missingness may not always be represented explicitly by `?`.

No preprocessing or missing-value handling decisions have been made yet. This stage is strictly concerned with understanding the raw site data before modifying it.


In [28]:
cleveland = cleveland.replace("?",np.nan)
hungary = hungary.replace("?",np.nan)
switzerland = switzerland.replace("?",np.nan)
va = va.replace("?", np.nan)

In [29]:
switzerland["chol"] = switzerland["chol"].replace(0, np.nan)
va["chol"] = va["chol"].replace(0,np.nan)

In [37]:
# Verifying that the cleaning of "?" and "0" worked correctly

print((cleveland == "?").sum())
print((hungary == "?").sum())
print((switzerland == "?").sum())
print((va == "?").sum())
print((switzerland["chol"] == 0).sum())
print((va["chol"] == 0).sum())


age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
0
0


In [43]:
print(cleveland.isna().sum())
print(hungary.isna().sum())
print(switzerland.isna().sum())
print(va.isna().sum())

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
num         0
dtype: int64
age           0
sex           0
cp            0
trestbps      1
chol         23
fbs           8
restecg       1
thalach       1
exang         1
oldpeak       0
slope       190
ca          291
thal        266
num           0
dtype: int64
age           0
sex           0
cp            0
trestbps      2
chol        123
fbs          75
restecg       1
thalach       1
exang         1
oldpeak       6
slope        17
ca          118
thal         52
num           0
dtype: int64
age           0
sex           0
cp            0
trestbps     56
chol          7
fbs           7
restecg       0
thalach      53
exang        53
oldpeak      56
slope       102
ca          198
thal        166
num           0
dtype: int64


###Stage 2: Understand the target

- original num values
- site-wise class balance
- binary target justified
- binary target created and verified

In [44]:
cleveland["num"].unique()

array([0, 2, 1, 3, 4])

In [47]:
hungary["num"].unique()

array([0, 1])

In [46]:
switzerland["num"].unique()

array([1, 3, 2, 0, 4])

In [48]:
va["num"].unique()

array([2, 0, 1, 3, 4])

Hungary only contains 0 and 1, while the other three contain 0–4, the concern that it raises if we tried to treat num as a 5-class target across all four sites is that we have no examples at all for classes 2, 3, and 4 in the hungary dataset.

That creates a serious cross-site comparability problem. A model evaluated on Hungary could never be tested on those classes there, because they simply do not exist in that site's labels.

Binary presence/absence gives all four sites a common, interpretable target and keeps the focus on cross-site generalisation, which is what we want to study.

A 5-class version would introduce a second research problem—unequal/missing class representation across sites—before we have even learned the centralized multi-site workflow.

Before recoding, we want to know the class distribution at each site:

0      → disease absent

1–4    → disease present

Why this matters: two sites can have the same target definition but very different proportions of positive and negative cases. That would be another form of site heterogeneity and could later affect both training and how we interpret metrics.

In [50]:
(cleveland["num"] == 0).sum()

np.int64(164)

So, out of 303 examples we have 164 as zero so (303 - 164) are 1- 4

In [51]:
(hungary["num"] == 0).sum()

np.int64(188)

Out of 294, 188 are zero.

In [52]:
(switzerland["num"] == 0).sum()

np.int64(8)

Out of 123, 8 are zero

In [53]:
(va["num"] == 0).sum()

np.int64(51)

Out of 200, 51 are 0.

Cleveland: 164 absent, 139 present

Hungary: 188 absent, 106 present

Switzerland: 8 absent, 115 present

VA Long Beach: 51 absent, 149 present


---

```
The proportions are very different across sites:

Cleveland: ~54% absent / 46% present

Hungary: ~64% absent / 36% present
Switzerland: ~7% absent / 93% present
VA: ~26% absent / 75% present

That is a major finding already: target distribution differs substantially by site.

```



In [63]:
cleveland["binary_target"] = (cleveland["num"]!= 0).astype(int)

In [64]:
hungary["binary_target"] = (hungary["num"]!= 0).astype(int)

In [65]:
switzerland["binary_target"] = (switzerland["num"] != 0 ).astype(int)

In [66]:
va["binary_target"] = (va["num"]!= 0).astype(int)

We have converted the target to:

0   → 0  (disease absent)

1–4 → 1  (disease present)

We preserve the original **num** column rather than overwrite it, and create a new binary target column. That way we retain the raw labels for traceability.

In [69]:
# Check if binary conversion for the target label worked
print(cleveland["binary_target"].unique())
print(switzerland["binary_target"].unique())
print(hungary["binary_target"].unique())
print(va["binary_target"].unique())

[0 1]
[0 1]
[0 1]
[0 1]
